plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java

```diff
@@ -1169,27 +1168,43 @@ private String getDataStoreUrlFromStore(DataStoreTO store) {
         }
         return store.getUrl();
     }
-
-    protected synchronized String attachOrDetachDevice(final Connect conn, final boolean attach, final String vmName, final String xml) throws LibvirtException, InternalErrorException {
+    protected synchronized void attachOrDetachDevice(final Connect conn, final boolean attach, final String vmName, final DiskDef xml)
+            throws LibvirtException, InternalErrorException {
+        attachOrDetachDevice(conn, attach, vmName, xml, 0l);
+    }
+    protected synchronized void attachOrDetachDevice(final Connect conn, final boolean attach, final String vmName, final DiskDef diskDef, long waitDetachDevice)
+            throws LibvirtException, InternalErrorException {
         Domain dm = null;
+        String diskXml = diskDef.toString();
+        String diskPath = diskDef.getDiskPath();
         try {
             dm = conn.domainLookupByName(vmName);
 
             if (attach) {
-                s_logger.debug("Attaching device: " + xml);
-                dm.attachDevice(xml);
-            } else {
-                s_logger.debug("Detaching device: " + xml);
-                dm.detachDevice(xml);
-                LibvirtDomainXMLParser parser = new LibvirtDomainXMLParser();
-                parser.parseDomainXML(dm.getXMLDesc(0));
-                List<DiskDef> disks = parser.getDisks();
-                for (DiskDef diskDef : disks) {
-                    if (StringUtils.contains(xml, diskDef.getDiskPath())) {
-                        throw new InternalErrorException("Could not detach volume. Probably the VM is in boot state at the moment");
-                    }
+                s_logger.debug("Attaching device: " + diskXml);
+                dm.attachDevice(diskXml);
+                return;
+            }
+            s_logger.debug(String.format("Detaching device: [%s].", diskXml));
+            dm.detachDevice(diskXml);
+            long wait = waitDetachDevice;
+            while (!checkDetachSucess(diskPath, dm) && wait > 0) {
+                try {
+                    wait -= waitDelayForVirshCommands;
+                    Thread.sleep(waitDelayForVirshCommands);
+                    s_logger.trace(String.format("Trying to detach device [%s] from VM instance with UUID [%s]. " +
+                            "Waiting [%s] milliseconds before assuming the VM was unable to detach the volume.", diskPath, dm.getUUIDString(), wait));
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
                 }
```

Comment:
extract this block into a method. Also a sonar suggestion but this one I very much agree on. It makes code both more readable and more debug-able. The exception can be cloudstack specific.

```suggestion
                try {
                    wait -= waitDelayForVirshCommands;
                    Thread.sleep(waitDelayForVirshCommands);
                    s_logger.trace(String.format("Trying to detach device [%s] from VM instance with UUID [%s]. " +
                            "Waiting [%s] milliseconds before assuming the VM was unable to detach the volume.", diskPath, dm.getUUIDString(), wait));
                } catch (InterruptedException e) {
                    throw new CloudRuntimeException(e);
                }
```